### ISSUE TITLE
Questions about the calculation of the loss 

### ISSUE THREAD

**Issue by ggordonhall** [zhuchen03/VIBNet#5]
Hi,

Thanks for releasing this implementation! I have been looking through this code and your paper (arXiv version) and there are a couple of things that are not clear to me:

1) `kl_mult` (an argument to each `InformationBottleneck` layer) corresponds to \gamma_i in equation (10). Then, after you sum the \gamma weighted KLDs of each IB layer you multiply the resulting `kl_total` by `kl_fac` before adding it to the cross entropy loss.
        
        ce_loss = criterion(output, target_var)
        
        loss = ce_loss
        if kl_fac > 0:
            loss += kl_total * kl_fac

    What is the motivation for `kl_fac`? It does not appear to derive from (10).

2) In the code I can't find the weighting term L on the data term in (10). Can you point me to where you scale this term? It occurred to me that `kl_fac` might be scaling down the KL term instead of scaling up the data term, but in that case I would expect the value of `kl_fac` to be 1/L, much larger than the default value of 1e-6.

Thanks again!

**Comment by daib13** [comment]
@ggordonhall Thanks for your interest to our work and sorry for the confusion.

1. Please check page 12 of the arxiv version about the scheme for \gamma_i. We choose \gamma_i = \gamma^\prime / S_i where S_i is the side length of the feature map. So `kl_mult' in the code is equivalent to 1/S_i. 

2. `kl_fac' is equivalent to \gamma^\prime / L. So the loss used in the code is (10) in the paper multiplied by 1/L.

I hope this can help you. 

Best

**Comment by ggordonhall** [comment]
Hi @daib13 

That definitely clears things up for (1), but I'm still confused about (2). 

In the code \gamma^\prime is 1 and L is 13-16 for the different architectures that you implement. If `kl_fac` = \gamma^\prime / L the value should be 1/16, which is much larger than the default `kl_fac` value of 1e-6. How do you arrive at this default value? Or do you just override it with a `kl_fac` argument closer to 1/16?

Thanks again!

**Comment by daib13** [comment]
Hi @ggordonhall ,

In the paper, \gamma^\prime is a hyper parameter to tune. In the code, we tune kl_fac instead, which is a scaled version of \gamma^\prime. So kl_fac = 1e-6 means \gamma^\prime = 1e-6 * L.

Best.

**Comment by ggordonhall** [comment]
Hi @daib13 

I see what you mean now. Thanks for the quick response.

Cheers
